axum-core/src/extract/mod.rs

```diff
@@ -65,256 +72,91 @@ pub use self::from_ref::FromRef;
 /// [`http::Request<B>`]: http::Request
 /// [`axum::extract`]: https://docs.rs/axum/latest/axum/extract/index.html
 #[async_trait]
-pub trait FromRequest<S, B>: Sized {
+pub trait FromRequestMut<S, B>: Sized {
     /// If the extractor fails it'll use this "rejection" type. A rejection is
     /// a kind of error that can be converted into a response.
     type Rejection: IntoResponse;
 
     /// Perform the extraction.
-    async fn from_request(req: &mut RequestParts<S, B>) -> Result<Self, Self::Rejection>;
+    async fn from_request_mut(req: &mut Request<B>, state: &S) -> Result<Self, Self::Rejection>;
```

Comment:
Really liking just passing the request here but its still possible to consume the body even with an `&mut Request<B>`:

```rust
#[async_trait]
impl<S, T, B> FromRequestMut<S, B> for Foo
where
    T: FromRequestMut<S, B>,
    B: http_body::Body + Send + Unpin,
    S: Send + Sync,
{
    type Rejection = Infallible;

    async fn from_request_mut(
        req: &mut Request<B>,
        state: &S,
    ) -> Result<Option<T>, Self::Rejection> {
        let pull_something_from_the_body = req.data().await;

        ...
    }
}
```

I think thats pretty unfortunate. I suppose we could do

```rust
async fn from_request_mut(parts: &mut http::request::Parts, state: &S) -> Result<Self, Self::Rejection>

// or maybe `from_request_parts`?
```

ie not even give you the body at all. That still allows you nuke the extensions, but maybe thats fine?

Reply:
Well, `Bytes` / `String` and friends all only implementing `FromRequestOnce` should already heavily discourage trying to extract the body in `FromRequestMut`, but I wouldn't mind passing `&mut http::request::Parts` either.

Reply:
Yeah it would definitely be shooting yourself in the foot

> but I wouldn't mind passing `&mut http::request::Parts` either

I think that would be interesting to explore. Could even call the traits `FromRequestParts` and `FromRequest` 🤔

Reply:
I like that, it mirrors `IntoResponse` and `IntoResponseParts`! I think being able to nuke extensions is fine by the way.

Reply:
Oh yeah thats true! Its all coming together 😃

I'll change it :)